### ISSUE TITLE
Blog Archive Widget Has MySQLcentric Date Logic

### ISSUE THREAD

**Issue by gordonbanderson** [silverstripe/silverstripe-blog#366]
The `DATE_FORMAT` function in the section following https://github.com/silverstripe/silverstripe-blog/blob/master/code/widgets/BlogArchiveWidget.php#L93 is MySQL only.

**Comment by tractorcow** [comment]
Yep, I think we should give it the same solution that we did for the archive action.

Since you're familiar with the API now, want to submit a patch? :P You go-getter you!

**Comment by gordonbanderson** [comment]
I'm not winning on this with Postgres.  I made the change using the `formattedDatetimeClause` method, see https://github.com/gordonbanderson/silverstripe-blog/blob/coverage2/code/widgets/BlogArchiveWidget.php , with corresponding test https://github.com/gordonbanderson/silverstripe-blog/blob/coverage2/tests/BlogArchiveWidgetTest.php

The test passes locally with MySQL and SQLite, but Postgres fails with the following:

```
[Warning] pg_query_params(): Query failed: ERROR: column "BlogPost_Live.PublishDate" must appear in the GROUP BY clause or be used in an aggregate function LINE 1: SELECT DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "Si... ^
```

If I add that column, I get a similar error:
```
[Warning] pg_query_params(): Query failed: ERROR: column "SiteTree_Live.ClassName" must appear in the GROUP BY clause or be used in an aggregate function LINE 1: ...DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "SiteTree_... ^
```
Adding the missing field and reloading each time eventually starts asking for SiteTree fields from third party modules, which is obviously not going to work as a generic solution.

The next option is to drop down to SQL level, but the method getBlogPosts() in Blog.php has an option to extend it, meaning I cannot simply use the ParentID to filter by.  For example the extension call to `updateGetBlogPosts` could remove all blog posts that have an ID that is a prime number (extreme example but not impossible).  There is of course also the relevant version to take into account, namely Stage or Live.

Options?  The only one I can see is to use 2 queries, the first calling getBlogPosts() on Blog, then using the IDs of the posts returned to use an IN clause for a second query that does the actual grouping by either year or year and month.  Feels icky though.

Fresh insight needed :)

**Comment by tractorcow** [comment]
Try adding it to the select fields as an expression alias, then grouping by that alias instead?

Also, please add the whole SQL query you're getting; I can't debug just from the error message.

**Comment by gordonbanderson** [comment]
Got sidetracked on on other things, now revisiting.

I've pushed changes, the last 3 commits, on https://github.com/gordonbanderson/silverstripe-blog/commits/issue_366_wip - a test for BlogArchiveWidget and updated code with respect to the dates.  Tests pass on SQLite and MySQL but not Postgres.  I need to rebase as my commit messages are lies when referring to Postgres :)

```bash
gordon@lapdance1404:~/work/git/weboftalent/sstest/www$ vendor/bin/phpunit blog/tests/BlogArchiveWidgetTest.php 
PHPUnit 3.7.38 by Sebastian Bergmann.

Configuration read from /home/gordon/work/git/weboftalent/sstest/www/phpunit.xml.dist

.EE

Time: 30.23 seconds, Memory: 48.25Mb

There were 2 errors:

1) BlogArchiveWidgetTest::testGetArchiveYearly
pg_query_params(): Query failed: ERROR:  column "BlogPost.PublishDate" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "Si...
                        ^

/home/gordon/work/git/weboftalent/sstest/www/postgresql/code/PostgreSQLConnector.php:200
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:134
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:178
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:136
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DB.php:311
/home/gordon/work/git/weboftalent/sstest/www/framework/model/queries/SQLExpression.php:121
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DataList.php:654
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DataList.php:748
/home/gordon/work/git/weboftalent/sstest/www/blog/code/widgets/BlogArchiveWidget.php:110
/home/gordon/work/git/weboftalent/sstest/www/blog/tests/BlogArchiveWidgetTest.php:33
/home/gordon/work/git/weboftalent/sstest/www/vendor/phpunit/phpunit/composer/bin/phpunit:63

2) BlogArchiveWidgetTest::testGetArchiveMonthly
pg_query_params(): Query failed: ERROR:  column "BlogPost.PublishDate" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "Si...
                        ^

/home/gordon/work/git/weboftalent/sstest/www/postgresql/code/PostgreSQLConnector.php:200
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:134
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:178
/home/gordon/work/git/weboftalent/sstest/www/framework/model/connect/Database.php:136
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DB.php:311
/home/gordon/work/git/weboftalent/sstest/www/framework/model/queries/SQLExpression.php:121
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DataList.php:654
/home/gordon/work/git/weboftalent/sstest/www/framework/model/DataList.php:748
/home/gordon/work/git/weboftalent/sstest/www/blog/code/widgets/BlogArchiveWidget.php:110
/home/gordon/work/git/weboftalent/sstest/www/blog/tests/BlogArchiveWidgetTest.php:68
/home/gordon/work/git/weboftalent/sstest/www/vendor/phpunit/phpunit/composer/bin/phpunit:63

FAILURES!
Tests: 3, Assertions: 1, Errors: 2.
```

**Comment by gordonbanderson** [comment]
Adding error_log debug when using Postgres the queries are:

```sql
SELECT DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "SiteTree"."ClassName", "SiteTree"."LastEdited", "SiteTree"."Created", "SiteTree"."URLSegment", "SiteTree"."Title", "SiteTree"."MenuTitle", "SiteTree"."Content", "SiteTree"."MetaDescription", "SiteTree"."ExtraMeta", "SiteTree"."ShowInMenus", "SiteTree"."ShowInSearch", "SiteTree"."Sort", "SiteTree"."HasBrokenFile", "SiteTree"."HasBrokenLink", "SiteTree"."ReportClass", "SiteTree"."CanViewType", "SiteTree"."CanEditType", "SiteTree"."Version", "SiteTree"."ParentID", "BlogPost"."PublishDate", "BlogPost"."AuthorNames", "BlogPost"."Summary", "BlogPost"."FeaturedImageID", "SiteTree"."ID", 
			CASE WHEN "SiteTree"."ClassName" IS NOT NULL THEN "SiteTree"."ClassName"
			ELSE 'SiteTree' END AS "RecordClassName"
 FROM "SiteTree" LEFT JOIN "BlogPost" ON "BlogPost"."ID" = "SiteTree"."ID"
 WHERE ("SiteTree"."ParentID" = ?)
 AND ("SiteTree"."ClassName" IN (?, ?))
 GROUP BY to_char("BlogPost"."PublishDate", TEXT 'YYYY')
 ORDER BY "_SortColumn0" DESC, "PublishDate" DESC

```sql
SELECT DISTINCT "PublishDate" IS NULL AS "_SortColumn0", "SiteTree"."ClassName", "SiteTree"."LastEdited", "SiteTree"."Created", "SiteTree"."URLSegment", "SiteTree"."Title", "SiteTree"."MenuTitle", "SiteTree"."Content", "SiteTree"."MetaDescription", "SiteTree"."ExtraMeta", "SiteTree"."ShowInMenus", "SiteTree"."ShowInSearch", "SiteTree"."Sort", "SiteTree"."HasBrokenFile", "SiteTree"."HasBrokenLink", "SiteTree"."ReportClass", "SiteTree"."CanViewType", "SiteTree"."CanEditType", "SiteTree"."Version", "SiteTree"."ParentID", "BlogPost"."PublishDate", "BlogPost"."AuthorNames", "BlogPost"."Summary", "BlogPost"."FeaturedImageID", "SiteTree"."ID", 
			CASE WHEN "SiteTree"."ClassName" IS NOT NULL THEN "SiteTree"."ClassName"
			ELSE 'SiteTree' END AS "RecordClassName"
 FROM "SiteTree" LEFT JOIN "BlogPost" ON "BlogPost"."ID" = "SiteTree"."ID"
 WHERE ("SiteTree"."ParentID" = ?)
 AND ("SiteTree"."ClassName" IN (?, ?))
 GROUP BY to_char("BlogPost"."PublishDate", TEXT 'YYYY-MM')
 ORDER BY "_SortColumn0" DESC, "PublishDate" DESC
```
The only difference is the GroupBY being by years or years and months

**Comment by tractorcow** [comment]
Wow, looking at that code again, I wonder if we should even be bothering to re-use the underlying query. Why not just make a new one that gets the publish dates, and skips all of the select / sorts in that query?

**Comment by gordonbanderson** [comment]
It seems thave having the date maths in the group by clause causes issues, testing at handcrafted SQL level.

This does not work:

```sql
SELECT "BlogPost" . "PublishDate"
FROM "BlogPost"
GROUP BY to_char("BlogPost" . "PublishDate", TEXT 'YYYY')
;
```

This does:
```sql
SELECT to_char("BlogPost" . "PublishDate", TEXT 'YYYY') as YEAR from "BlogPost" GROUP BY YEAR;
```

**Comment by gordonbanderson** [comment]
The commits I've just pushed to https://github.com/gordonbanderson/silverstripe-blog/commits/issue_366_wip are passing locally on Postgres and on another branch different to this one on Travis for each of Postgres, MySQL and SQLite.

This solution uses a RAW SQL query, and note that it does not take account of

```php
        $this->extend('updateGetBlogPosts', $blogPosts);
```

in Blog::getBlogPosts() method.  I'm not sure if there is a 'clean' way to get the alterations to the query introduced optionally by that method.

I guess also values need escaped for SQL injection purposes, but can you check if this is a valid approach please @tractorcow even if needing fine tuned.

**Comment by tractorcow** [comment]
So nothing above https://github.com/gordonbanderson/silverstripe-blog/blob/cc0cdd86d19961fc567a57d89be6689f6f9fb261/code/widgets/BlogArchiveWidget.php#L103 is used?

Make sure at https://github.com/gordonbanderson/silverstripe-blog/blob/cc0cdd86d19961fc567a57d89be6689f6f9fb261/code/widgets/BlogArchiveWidget.php#L107 you don't add _Stage to each table if in draft mode.

You can put all of the raw query generation into an SQLQuery object.

**Comment by dhensby** [comment]
is this fixed?

**Comment by tractorcow** [comment]
No still broken.

**Comment by gordonbanderson** [comment]
Apologies I've not had time to look into it, life issues getting in the way at the moment.

**Comment by micmania1** [comment]
Can anybody summarise what still needs to be done to fix this?

**Comment by tractorcow** [comment]
Someone needs to swap out the code in https://github.com/silverstripe/silverstripe-blog/blob/fc6c86fd35843bf0e2581b5c8c46d16ca8978e09/code/widgets/BlogArchiveWidget.php#L93 with sql-agnostic code.

This is the correct way to get a year filter, for example,

```php
$timeConstraint = $conn->formattedDatetimeClause('"BlogPost"."PublishDate"', '%Y');
```

**Comment by lekoala** [comment]
by the way, the current group by seems to be broken on mysql 5.7 if ONLY_FULL_GROUP_BY is enabled. I hope switching to sql-agnostic code resolve this issue as well

**Comment by tractorcow** [comment]
Ah yes, that config is great for spotting poor sql design in core. ;)

**Comment by kinglozzer** [comment]
Yeah I’m also hitting the `ONLY_FULL_GROUP_BY` issue, and I can’t think of a workaround for it... maybe we should just do an `SQLQuery` so we’re only selecting the fields we need for the `GROUP BY` statement?

**Comment by kinglozzer** [comment]
Okay, this is what I’m using for my project (this is currently dumped in an extension) to get around the `ONLY_FULL_GROUP_BY` issue in MySQL. Perhaps this could be used as the basis for refactoring the widget method?

```php
/**
 * Returns a list of months where blog posts are present
 *
 * @return ArrayList
 */
public function getArchiveMonths()
{
    $stage = Versioned::current_stage();
    if ($stage === 'Stage' && !Permission::check('VIEW_DRAFT_CONTENT')) {
        $suffix = '';
    } else {
        $suffix = "_{$stage}";
    }

    $query = SQLSelect::create()
        ->setSelect([
            'FormattedMonth' => DB::get_conn()->formattedDatetimeClause('"PublishDate"', '%Y-%m'),
            'PublishDate' => 'MAX(PublishDate)'
        ])
        ->setFrom("BlogPost{$suffix}")
        ->setOrderBy('PublishDate DESC')
        ->setGroupBy('FormattedMonth');

    $archive = ArrayList::create();
    foreach ($query->execute() as $result) {
        $date = DBField::create_field('Date', $result['PublishDate']);

        $year = $date->Format('Y');
        $month = $date->Format('m');
        $title = $date->FormatI18N('%B %Y');

        $archive->push(ArrayData::create([
            'Title' => $title,
            'Link' => Controller::join_links($this->owner->Link('archive'), $year, $month, '/')
        ]));
    }

    return $archive;
}
```

**Comment by kinglozzer** [comment]
I found this link helpful in explaining the `GROUP BY` issues: http://rpbouman.blogspot.co.uk/2007/05/debunking-group-by-myths.html. Might explain why PGSQL is _always_ failing on the group by statements, and why MySQL is _sometimes_ failing (when `ONLY_FULL_GROUP_BY` is set)

**Comment by robbieaverill** [comment]
Fixed in #480 which has tests for Postgres and SQLite
